File: src/simulation/sph.py
"""
src/simulation/sph.py
=====================
2-D Smoothed Particle Hydrodynamics (SPH) solver using Taichi.

Mathematical kernels (Müller et al. 2003):
  • Poly6    — density estimation       W(r,h)
  • Spiky    — pressure gradient        ∇W(r,h)  (no vanishing centre gradient)
  • Viscosity Laplacian — vel diffusion ∇²W(r,h)

The outer particle loop is parallelised by Taichi (CPU SIMD or GPU).
The inner neighbour loop is sequential — O(n²) total but fast in practice
because ~99 % of pairs are culled by the r < h distance check.
"""

import math
import taichi as ti
import numpy as np


@ti.data_oriented
class SPHSolver:
    """
    GPU/CPU-parallel 2-D SPH fluid solver.

    Key public methods
    ------------------
    initialize()             — scatter particles, reset velocities
    step(substeps)           — advance simulation by `substeps` time steps
    set_force_sources(list)  — update (x,y,strength,radius) external forces
    set_gravity_tilt(angle)  — tilt gravity direction (head pose)
    get_render_data()        — return (positions, temperatures) as numpy arrays
    """

    def __init__(
        self,
        n:        int,
        width:    float,
        height:   float,
        h:        float  = 26.0,
        rho0:     float  = 280.0,
        k:        float  = 200.0,
        mu:       float  = 0.55,
        gravity:  float  = 200.0,
        dt:       float  = 0.008,
        mass:     float  = 1.0,
        damp:     float  = 0.4,
        max_src:  int    = 32,
    ):
        self.n      = n
        self.W      = width
        self.H      = height
        self.h      = h
        self.h2     = h * h
        self.rho0   = rho0
        self.k      = k
        self.mu     = mu
        self._gravity = gravity   # base magnitude
        self.dt     = dt
        self.mass   = mass
        self.damp   = damp
        self.max_src = max_src

        # Kernel normalisation constants (2-D)
        pi = math.pi
        self.c_poly6 =  4.0  / (pi * h ** 8)
        self.c_spiky = -30.0 / (pi * h ** 5)
        self.c_visc  =  40.0 / (pi * h ** 5)

        # ── Particle state ───────────────────────────────────────────────────
        self.pos  = ti.Vector.field(2, ti.f32, n)
        self.vel  = ti.Vector.field(2, ti.f32, n)
        self.rho  = ti.field(ti.f32, n)
        self.pres = ti.field(ti.f32, n)
        self.temp = ti.field(ti.f32, n)   # visual heat [0, 1]

        # ── External force sources ───────────────────────────────────────────
        self.n_src   = ti.field(ti.i32, shape=())
        self.src_pos = ti.Vector.field(2, ti.f32, max_src)
        self.src_str = ti.field(ti.f32, max_src)   # +attract / −repel
        self.src_rad = ti.field(ti.f32, max_src)

        # ── Gravity (can be tilted via head pose) ────────────────────────────
        self.grav_x = ti.field(ti.f32, shape=())
        self.grav_y = ti.field(ti.f32, shape=())

    # ─────────────────────────────────────────────────────────────────────────
    # Initialisation
    # ─────────────────────────────────────────────────────────────────────────

    def initialize(self):
        """Reset simulation: scatter particles uniformly, zero velocities."""
        self.grav_x[None] = 0.0
        self.grav_y[None] = float(self._gravity)
        self.n_src[None]  = 0
        self._init_kernel()

    @ti.kernel
    def _init_kernel(self):
        for i in self.pos:
            self.pos[i]  = ti.Vector([
                ti.random() * self.W,
                ti.random() * self.H * 0.8 + self.H * 0.1,
            ])
            self.vel[i]  = ti.Vector([0.0, 0.0])
            self.temp[i] = ti.random() * 0.15

    # ─────────────────────────────────────────────────────────────────────────
    # Kernel functions  (called inside @ti.kernel, inline-expanded by Taichi)
    # ─────────────────────────────────────────────────────────────────────────

    @ti.func
    def _poly6(self, r2: ti.f32) -> ti.f32:
        v = 0.0
        if r2 < self.h2:
            d = self.h2 - r2
            v = self.c_poly6 * d * d * d
        return v

    @ti.func
    def _spiky_grad(self, r_vec: ti.template()) -> ti.template():
        r = r_vec.norm()
        g = ti.Vector([0.0, 0.0])
        if r > 1e-4 and r < self.h:
            d = self.h - r
            g = self.c_spiky * d * d * (r_vec / r)
        return g

    @ti.func
    def _visc_lap(self, r: ti.f32) -> ti.f32:
        v = 0.0
        if r < self.h:
            v = self.c_visc * (self.h - r)
        return v

    # ─────────────────────────────────────────────────────────────────────────
    # Physics kernels
    # ─────────────────────────────────────────────────────────────────────────

    @ti.kernel
    def _compute_density_pressure(self):
        for i in range(self.n):
            rho_i = 0.0
            for j in range(self.n):
                r_vec = self.pos[i] - self.pos[j]
                rho_i += self.mass * self._poly6(r_vec.dot(r_vec))
            self.rho[i]  = rho_i
            # Tait equation of state (clamped to avoid tensile instability)
            self.pres[i] = ti.max(self.k * (rho_i - self.rho0), 0.0)

    @ti.kernel
    def _compute_forces_integrate(self):
        n_s = self.n_src[None]

        for i in range(self.n):
            f_pres = ti.Vector([0.0, 0.0])
            f_visc = ti.Vector([0.0, 0.0])
            rho_i  = ti.max(self.rho[i], 1e-3)
            pres_i = self.pres[i]
            vel_i  = self.vel[i]
            pos_i  = self.pos[i]

            for j in range(self.n):
                if i == j:
                    continue
                r_vec  = pos_i - self.pos[j]
                r      = r_vec.norm()
                rho_j  = ti.max(self.rho[j], 1e-3)

                if r < self.h and r > 1e-4:
                    # Symmetric pressure (conserves momentum exactly)
                    f_pres += (
                        -self.mass * (pres_i + self.pres[j]) / (2.0 * rho_j)
                        * self._spiky_grad(r_vec)
                    )
                    # Viscosity
                    f_visc += (
                        self.mass * self.mu * (self.vel[j] - vel_i)
                        / rho_j * self._visc_lap(r)
                    )

            # Gravity (direction may be tilted by head pose)
            f_grav = ti.Vector([self.grav_x[None], self.grav_y[None]]) * rho_i

            # External force sources (hands / face)
            f_ext      = ti.Vector([0.0, 0.0])
            heat_delta = 0.0
            for s in range(n_s):
                d_vec  = self.src_pos[s] - pos_i
                d      = d_vec.norm()
                r_src  = self.src_rad[s]
                if d < r_src and d > 1e-4:
                    w       = (r_src - d) / r_src          # 1 at source, 0 at edge
                    f_ext  += self.src_str[s] * w * (d_vec / d) * rho_i
                    heat_delta = ti.max(heat_delta, w * 0.12)

            # ── Semi-implicit Euler integration ───────────────────────────────
            acc     = (f_pres + f_visc + f_grav + f_ext) / rho_i
            new_vel = vel_i + acc * self.dt
            new_pos = pos_i + new_vel * self.dt

            # ── Boundary conditions ───────────────────────────────────────────
            br = 2.0   # particle boundary offset
            if new_pos[0] < br:
                new_pos[0] = br
                new_vel[0] = ti.abs(new_vel[0]) * self.damp
            if new_pos[0] > self.W - br:
                new_pos[0] = self.W - br
                new_vel[0] = -ti.abs(new_vel[0]) * self.damp
            if new_pos[1] < br:
                new_pos[1] = br
                new_vel[1] = ti.abs(new_vel[1]) * self.damp
            if new_pos[1] > self.H - br:
                new_pos[1] = self.H - br
                new_vel[1] = -ti.abs(new_vel[1]) * self.damp

            self.vel[i] = new_vel
            self.pos[i] = new_pos

            # ── Visual temperature (drives particle colour) ───────────────────
            speed  = new_vel.norm()
            target = ti.min(speed * 0.004 + heat_delta, 1.0)
            self.temp[i] = self.temp[i] * 0.94 + target * 0.06

    # ─────────────────────────────────────────────────────────────────────────
    # Public API
    # ─────────────────────────────────────────────────────────────────────────

    def step(self, substeps: int = 3):
        """Advance simulation by `substeps` time steps."""
        for _ in range(substeps):
            self._compute_density_pressure()
            self._compute_forces_integrate()

    def set_force_sources(self, sources: list):
        """
        sources : list of (x, y, strength, radius) tuples.
          strength > 0  → attract particles toward source
          strength < 0  → repel particles away from source
        """
        n = min(len(sources), self.max_src)
        self.n_src[None] = n
        for idx, (x, y, s, rad) in enumerate(sources[:n]):
            self.src_pos[idx] = [float(x), float(y)]
            self.src_str[idx] = float(s)
            self.src_rad[idx] = float(rad)

    def set_gravity_tilt(self, angle_rad: float):
        """Tilt gravity direction by `angle_rad` radians from vertical."""
        g = float(self._gravity)
        self.grav_x[None] = math.sin(angle_rad) * g
        self.grav_y[None] = math.cos(angle_rad) * g

    def disable_gravity(self):
        self.grav_x[None] = 0.0
        self.grav_y[None] = 0.0

    def enable_gravity(self):
        self.grav_x[None] = 0.0
        self.grav_y[None] = float(self._gravity)

    def get_render_data(self):
        """Return (positions, temperatures) as numpy float32 arrays."""
        return self.pos.to_numpy(), self.temp.to_numpy()

File: lib/sph.ts
/**
 * 2-D SPH fluid solver (TypeScript).
 *
 * Spatial hashing for O(n * avg_neighbours) neighbour lookup.
 * Supports vortex forces (perpendicular to radial direction).
 */

const PI = Math.PI;

export interface ForceSource {
  x: number;
  y: number;
  strength: number;
  radius: number;
  vortex?: number; // perpendicular force component (spin)
}

export class SPHSolver {
  readonly n: number;
  readonly W: number;
  readonly H: number;

  // SoA particle state
  px: Float32Array;
  py: Float32Array;
  vx: Float32Array;
  vy: Float32Array;
  rho: Float32Array;
  pres: Float32Array;
  temp: Float32Array;

  // Spatial hash grid
  private cellSize: number;
  private gridW: number;
  private gridH: number;
  private nCells: number;
  private cellCount: Int32Array;
  private cellStart: Int32Array;
  private order: Int32Array;
  private writePos: Int32Array;

  // Physics
  private h: number;
  private h2: number;
  private rho0: number;
  private k: number;
  mu: number;
  private dt: number;
  private mass: number;
  private damp: number;
  private cPoly6: number;
  private cSpiky: number;
  private cVisc: number;

  gravX = 0;
  gravY: number;
  sources: ForceSource[] = [];

  constructor(n: number, width: number, height: number, h = 26, rho0 = 280, k = 200, mu = 0.55, gravity = 200, dt = 0.008, mass = 1.0, damp = 0.4) {
    this.n = n;
    this.W = width;
    this.H = height;
    this.h = h;
    this.h2 = h * h;
    this.rho0 = rho0;
    this.k = k;
    this.mu = mu;
    this.dt = dt;
    this.mass = mass;
    this.damp = damp;
    this.gravY = gravity;
    this.cPoly6 = 4.0 / (PI * h ** 8);
    this.cSpiky = -30.0 / (PI * h ** 5);
    this.cVisc = 40.0 / (PI * h ** 5);

    this.px = new Float32Array(n);
    this.py = new Float32Array(n);
    this.vx = new Float32Array(n);
    this.vy = new Float32Array(n);
    this.rho = new Float32Array(n);
    this.pres = new Float32Array(n);
    this.temp = new Float32Array(n);

    this.cellSize = h;
    this.gridW = Math.ceil(width / h) + 1;
    this.gridH = Math.ceil(height / h) + 1;
    this.nCells = this.gridW * this.gridH;
    this.cellCount = new Int32Array(this.nCells);
    this.cellStart = new Int32Array(this.nCells);
    this.order = new Int32Array(n);
    this.writePos = new Int32Array(this.nCells);
  }

  initialize(): void {
    for (let i = 0; i < this.n; i++) {
      this.px[i] = Math.random() * this.W;
      this.py[i] = Math.random() * this.H * 0.8 + this.H * 0.1;
      this.vx[i] = 0;
      this.vy[i] = 0;
      this.temp[i] = Math.random() * 0.15;
    }
    this.gravX = 0;
    this.gravY = 200;
  }

  setGravityTilt(angle: number): void {
    const g = Math.sqrt(this.gravX ** 2 + this.gravY ** 2) || 200;
    this.gravX = Math.sin(angle) * g;
    this.gravY = Math.cos(angle) * g;
  }

  // ── Spatial hash ─────────────────────────────────────────────────────────

  private buildGrid(): void {
    this.cellCount.fill(0);
    const cs = this.cellSize;
    const gw = this.gridW;
    const gh = this.gridH;
    for (let i = 0; i < this.n; i++) {
      const ci = Math.min(Math.max(0, (this.px[i] / cs) | 0), gw - 1);
      const cj = Math.min(Math.max(0, (this.py[i] / cs) | 0), gh - 1);
      this.cellCount[ci + cj * gw]++;
    }
    let sum = 0;
    for (let c = 0; c < this.nCells; c++) {
      this.cellStart[c] = sum;
      sum += this.cellCount[c];
    }
    this.writePos.set(this.cellStart);
    for (let i = 0; i < this.n; i++) {
      const ci = Math.min(Math.max(0, (this.px[i] / cs) | 0), gw - 1);
      const cj = Math.min(Math.max(0, (this.py[i] / cs) | 0), gh - 1);
      const idx = ci + cj * gw;
      this.order[this.writePos[idx]++] = i;
    }
  }

  // ── SPH kernels ──────────────────────────────────────────────────────────

  private computeDensityPressure(): void {
    const { n, h2, mass, rho0, k: kk, cPoly6 } = this;
    const cs = this.cellSize;
    const gw = this.gridW;
    const gh = this.gridH;
    const { px, py, rho, pres, cellStart, cellCount, order } = this;

    for (let i = 0; i < n; i++) {
      let rho_i = 0;
      const xi = px[i];
      const yi = py[i];
      const ci = Math.min(Math.max(0, (xi / cs) | 0), gw - 1);
      const cj = Math.min(Math.max(0, (yi / cs) | 0), gh - 1);
      for (let di = -1; di <= 1; di++) {
        const ni = ci + di;
        if (ni < 0 || ni >= gw) continue;
        for (let dj = -1; dj <= 1; dj++) {
          const nj = cj + dj;
          if (nj < 0 || nj >= gh) continue;
          const cell = ni + nj * gw;
          const start = cellStart[cell];
          const count = cellCount[cell];
          for (let idx = start; idx < start + count; idx++) {
            const j = order[idx];
            const dx = xi - px[j];
            const dy = yi - py[j];
            const r2 = dx * dx + dy * dy;
            if (r2 < h2) {
              const d = h2 - r2;
              rho_i += mass * cPoly6 * d * d * d;
            }
          }
        }
      }
      rho[i] = rho_i;
      pres[i] = Math.max(kk * (rho_i - rho0), 0);
    }
  }

  private computeForcesIntegrate(): void {
    const { n, h, h2, mass, mu, dt, damp, cSpiky, cVisc, px, py, vx, vy, rho, pres, sources } = this;
    const cs = this.cellSize;
    const gw = this.gridW;
    const gh = this.gridH;
    const { cellStart, cellCount, order } = this;
    const W = this.W;
    const H = this.H;
    const gravX = this.gravX;
    const gravY = this.gravY;

    for (let i = 0; i < n; i++) {
      let fpx = 0, fpy = 0, fvx = 0, fvy = 0;
      const xi = px[i], yi = py[i], vxi = vx[i], vyi = vy[i];
      const rho_i = Math.max(rho[i], 1e-3);
      const pres_i = pres[i];
      const ci = Math.min(Math.max(0, (xi / cs) | 0), gw - 1);
      const cj = Math.min(Math.max(0, (yi / cs) | 0), gh - 1);

      for (let di = -1; di <= 1; di++) {
        const ni = ci + di;
        if (ni < 0 || ni >= gw) continue;
        for (let dj = -1; dj <= 1; dj++) {
          const nj = cj + dj;
          if (nj < 0 || nj >= gh) continue;
          const cell = ni + nj * gw;
          const start = cellStart[cell];
          const count = cellCount[cell];
          for (let idx = start; idx < start + count; idx++) {
            const j = order[idx];
            if (i === j) continue;
            const dx = xi - px[j];
            const dy = yi - py[j];
            const r2 = dx * dx + dy * dy;
            if (r2 >= h2 || r2 < 1e-8) continue;
            const r = Math.sqrt(r2);
            const rho_j = Math.max(rho[j], 1e-3);
            const d = h - r;
            const scale = -mass * (pres_i + pres[j]) / (2 * rho_j) * cSpiky * d * d;
            const invR = 1 / r;
            fpx += scale * dx * invR;
            fpy += scale * dy * invR;
            const viscScale = mass * mu / rho_j * cVisc * (h - r);
            fvx += viscScale * (vx[j] - vxi);
            fvy += viscScale * (vy[j] - vyi);
          }
        }
      }

      const fgx = gravX * rho_i;
      const fgy = gravY * rho_i;

      // External forces (with vortex support)
      let fex = 0, fey = 0, heatDelta = 0;
      for (let s = 0; s < sources.length; s++) {
        const src = sources[s];
        const sdx = src.x - xi;
        const sdy = src.y - yi;
        const sd = Math.sqrt(sdx * sdx + sdy * sdy);
        if (sd < src.radius && sd > 1e-4) {
          const w = (src.radius - sd) / src.radius;
          const invD = 1 / sd;
          // Radial force
          fex += src.strength * w * sdx * invD * rho_i;
          fey += src.strength * w * sdy * invD * rho_i;
          // Vortex (perpendicular) force
          if (src.vortex) {
            fex += src.vortex * w * (-sdy * invD) * rho_i;
            fey += src.vortex * w * (sdx * invD) * rho_i;
          }
          heatDelta = Math.max(heatDelta, w * 0.12);
        }
      }

      const invRho = 1 / rho_i;
      const ax = (fpx + fvx + fgx + fex) * invRho;
      const ay = (fpy + fvy + fgy + fey) * invRho;
      let nvx = vxi + ax * dt;
      let nvy = vyi + ay * dt;
      let npx = xi + nvx * dt;
      let npy = yi + nvy * dt;

      const br = 2;
      if (npx < br) { npx = br; nvx = Math.abs(nvx) * damp; }
      if (npx > W - br) { npx = W - br; nvx = -Math.abs(nvx) * damp; }
      if (npy < br) { npy = br; nvy = Math.abs(nvy) * damp; }
      if (npy > H - br) { npy = H - br; nvy = -Math.abs(nvy) * damp; }

      vx[i] = nvx;
      vy[i] = nvy;
      px[i] = npx;
      py[i] = npy;

      const speed = Math.sqrt(nvx * nvx + nvy * nvy);
      const target = Math.min(speed * 0.004 + heatDelta, 1.0);
      this.temp[i] = this.temp[i] * 0.94 + target * 0.06;
    }
  }

  step(substeps = 3): void {
    for (let s = 0; s < substeps; s++) {
      this.buildGrid();
      this.computeDensityPressure();
      this.computeForcesIntegrate();
    }
  }
}
